CONTRIBUTING.md

```diff
@@ -0,0 +1,417 @@
+  1. Main Module Header (i.e., the interface that we implement in a given .cpp)
+  2. Local/Private Headers
+  3. MLIR project/subproject headers (as separate block)
+  4. LLVM project/subproject headers (`clang/...`, `lldb/...`, `llvm/...`, etc, except MLIR)
+  5. System #includes
+
+### Debug Output
+
+Note that `llvm::outs()` is not thread-safe (since it is buffered) and thus cannot be used in code that will run multi-threaded. For short-lived debug messages, printing to `llvm:errs()` is fine (also thread-safe). If debug messages should remain in the code, consider using `LLVM_DEBUG` (see [the LLVM Programmer's Reference Manual - Section The LLVM_DEBUG() macro and -debug option](https://llvm.org/docs/ProgrammersManual.html#the-llvm-debug-macro-and-debug-option)).
+
+### Static Code Checks
+The easiest, fastest, and most automated way to integrate the formatting into your workflow
+is via [pre-commit](https://pre-commit.com). Note that this tool requires and internet connection
+to initially setup because the formatting tools needs to be downloaded.
+
+**In environments without an internet connection, please see one of the other solutions documented
+below.**
+
+These should be installed and setup prior to any development work so that they are not forgotten
+about. The setup is straight forward:
+
+```bash
+pip install pre-commit
+pre-commit install
+```
+
+The first time `pre-commit install` is run, it will take a minute to setup the environment.
+
+After installation, the hooks will be run prior to every commit, and will be run against all staged
+changes. Optionally, you can trigger this run via `pre-commit run`.
+
+If you wish to run the hooks against the entire repository, run:
+
+```bash
+pre-commit run --all-files
+```
+
+The other checks that are performed can be seen in
+[`.pre-commit-config.yaml`](.pre-commit-config.yaml). At the time of writing, these are:
+
+- No direct committing to `main`, or `release/*`
+- Check json for validity
+- Ensure newline character at the end of files
+- Trim end of line whitespace characters
+- Check for no merge conflict lines accidentally being staged
+- Clang format
+- Python block (line length 100)
+
+
+## Testing and Analysis
+
+Several targets and CMake configuration options are defined as shown below to test the qe-compiler. The standard test suite consists of a format check to ensure the source follows the configured style defined in `.clang-format`, a run of `clang-tidy` to perform code analysis, and a run of the test suite. It is also possible to run additional analysis with Valgrind and the Google Sanitizers below. CMake commands should always be run from a build directory.
+
+Our test suite contains both unit tests (using googletest) and integration tests that exercise the full qss-compiler or qss-opt tools (using LLVM lit).
```

Comment:
```suggestion
Our test suite contains both unit tests (using googletest) and integration tests that exercise the full qe-compiler or qss-opt tools (using LLVM lit).
```

Reply:
What is `qss-opt`?

Reply:
I kept this the same as the binaries are still named qss-compiler.